### ISSUE TITLE
Support for value interpolation

### ISSUE THREAD

**Issue by brandonmcconnell** [Noriller/easy-tailwind#2]
I recently began work on a Tailwind plugin different from yours in that mine did not support grouping or anything like clsx, but similar in that mine introduces a transformer into the Tailwind config, so that I can inject constructed utility names without the entire class name needing to be fully written out.

The goal was to support value interpolation, but it appears your plugin may already cover this too. I just wanted to check to verify that.

My usage looks more like this:
```js
const percent = 27;
const className = twerp(`content-[${percent}%]`, range(0, 100));
```

Here, a value can be dynamically interpolated using `twerp` but with all possible interpolatable values please as the second arg, with an array of arrays when multiple interpolatable values are present.

Is something like this already possible and natively supported using Easy Tailwind?

**Comment by Noriller** [comment]
Not that I'm aware of.
The problem, as you know, is that you need to inject all of that into tailwind.
What I do here is basically recalculate everything at that step, only dropping conditionals to generate all classes to be used.
One thing It's not really possible to support without much pain is variables from elsewhere.

Considering your scope, if I had to do it, I would possibly use regex to find all mentions in the code of `twerp`, change the string interpolation with some random string and have a function that looks for that and replaces that with the values from the range returning all possible classes. Not sure if it would be the best approach for this case though.

**Comment by brandonmcconnell** [comment]
@Noriller Thanks for the feedback!

Another idea—and this one would likely take a bit more effort, especially to make it cross-framework compatible—is to interpolate CSS variables instead of actual values, which would drastically lighten the load for Tailwind.

However, if I wanted to support dynamically changing values which would probably be essential here, I'd need to inject inline styling for the CSS car value itself, since handling that value in tailwind would be just as complicated.

In Svelte-land, this would probably end up looking more like a custom rule like…

```svelte
<Component use:twerp="content-[{val}]" />
```

This rule could inject the tailwind utilities into the `class` attribute (and the TW config) and also inject the variable value into the element itself via inline styling.

I'll have to see if there's a way of doing something similar and React and Vue. There's probably a fairly straightforward way in both. In react, it'll probably require some use of useRef to manage the DOM mode itself.

Definitely open to more ideas or feedback if you have any. Thanks again! 🙌🏼

**Comment by Noriller** [comment]
Actually, you could just use vanilla JS to do that.

Maybe something like this: https://stackoverflow.com/questions/37801882/how-to-change-css-root-color-variables-in-javascript

the only problem would be to be able to use it multiple times without conflict... but by going with vanilla you wouldn't need to care about frameworks. (in this case, you would probably still need to inject something into tailwind.

**Comment by brandonmcconnell** [comment]
Yeah, the impl could be done in vanilla JS, though to make it easy to use in different frameworks, I think some framework-specific wrappers would be needed.

Maybe I'm thinking about this wrong, though. Theoretically, if I build this in vanilla JS and someone wants to use it a React codebase, like below:

```jsx
const RenderStringAsCSSContent = ({ str }) => {
  return <div className={`content-[${str}]`}></div>;
}
```

In this case, what would handle the creation of the inline style for the CSS var? Ideally, this would render to this:

```html
<div class="content-[var(--twerp-64cht67tb4)]" style="--twerp-64cht67tb4: 'Joe'"></div>
```

Or are saying that this could be done manually without any need for twerp, like this:

```jsx
import { useId } from 'react';

const RenderStringAsCSSContent = ({ str }) => {
  const twerpId = `--twerp-${useId()}`;
  return <div style={{ [twerpId]: str }} className={`content-[var(${str})]`}></div>;
}
```

This could certainly work, and more or less, I'm really only seeking to find a way to streamline this, so we can essentially write it as simply as `<Component class={twerp(`content-[${val}]`)} />` (or similar).

As you said, I'll likely still need to inject something into Tailwind, which I'm okay with. I'd love to find a way to make my transformer for `twerp` compatible with yours, or at least not conflicting, so both tools can be used in the same projects by Tailwind CSS users who really want to maximize their productivity. 🙂

**Comment by Noriller** [comment]
From what I understand, it has to create the classes and inject them, with variable names and all.
Being able to sync between renders and SSR would probably be a must, so it wouldn't be a trivial thing to do.

You could also lookup on how actual plugins work and do something like `content-[{0,${percent},100}]` and have it create all possible values for that. Not sure how the interpolate would work in this case though...

**Comment by brandonmcconnell** [comment]
Yeah, I'll give it some more thought and see what's possible. Thanks for thinking through all this with me!

Something like this might be less "nice" but easier for the parser to deal with, since it'll break up the strings from the dynamic bits:

```jsx
<Component className={twerp([`sm:w-[`, range(0, 100), `px]`])} />
```

or something like this pseudo-code:

```jsx
function twerp(prefix, values, suffix = '') {
  return values.map(v => `${prefix}-[${v}${suffix}]`).join(' ');
}

<Component className={twerp('sm:w', range(0, 100), 'px')} />
```

**Comment by Noriller** [comment]
That part is not really a problem.
Maybe you can try something like this:
https://github.com/Mokshit06/tw-variants

Here is a babel plugin, so you could expand on whats needed and still keep the interpolation in the finished code.
